### PR TITLE
Empty array support

### PR DESCRIPTION
**Pull request by manish-sethi**
This PR allows storing an empty array as a value of a key in the ledger. However, an attempt to store a 'nil' would result in an error - see issue# 936
Signed-off-by:manishsethi@in.ibm.com

### CODE REVIEW

**Comment by corecode** [comment]
Commit 715b2ca095e7b651da7aabd39044a0657e627992 is a mechanical pass over several files and should be labeled as such.  This commit is also independent of the contents of the PR.

File: core/ledger/ledger.go [hunk 1]
```diff
@@ -207,6 +207,9 @@ func (ledger *Ledger) GetStateRangeScanIterator(chaincodeID string, startKey str
 
 // SetState sets state to given value for chaincodeID and key. Does not immideatly writes to DB
 func (ledger *Ledger) SetState(chaincodeID string, key string, value []byte) error {
+	if key == "" || value == nil {
```
Comment (corecode) [inline core/ledger/ledger.go:240]: why is a nil key or value not supported?

**Comment by manish-sethi** [comment]
@corecode, when I ran goimports etc. tools, it picked up existing files also. May be, I should have run these tools only on selective files that were originally changed in this PR. Is that how you suggest?

File: core/ledger/ledger.go (hunk 1 above)
Comment (manish-sethi) [inline core/ledger/ledger.go:240]: This was never supported and tested. I have added the check explicitly here to avoid issues down stream. We can add support for these later but that requires changes down stream.

**Comment by corecode** [comment]
I think somebody should run goimports for the whole repo, do one monster commit (making sure that chaincode hashes are fixed appropriately), and then have CI to fail if commits have not been passed through goimports.

**Comment by binhn** [comment]
@corecode we probably should take goimports related discussion on another issue. A lot of problems reported are on public function comments, which I think would be more appropriate for the authors to do.

What are the use-cases for empty or nil keys?

**Comment by mastersingh24** [comment]
I agree with @binhn  - why do we need empty or nil keys and/or valuies?

**Comment by bcbrock** [comment]
I've made the case for empty values in #936. As for empty keys: The key is specified to be a string, "" is a unique string, so it should be accepted as a key. From a specification point of view there is no reason to disallow it; there is no ambiguity. As far as I can tell RocksDB allows the empty string as a key. Also compare specifications of other modern key/value stores like Redis

```
Redis keys are binary safe, this means that you can use any binary sequence as a key, 
from a string like "foo" to the content of a JPEG file. The empty string is also a valid key.
```

I've seen examples where keys are stored as fields of tables. If a field is "" it indicates a default value, which is represented by a key/value pair ""/<default>.

I think that `PutState` from Go should allow empty keys, empty values and `nil` values (see #936), so I think @corecode comment above is correct, and the test at `ledger.go:210` should be eliminated.

I don't understand the hesitation about downstream changes. Why wait to fix issues related to handling empty strings as keys? Why not fix them now?

**Comment by manish-sethi** [comment]
@bcbrock - For making the changes, we all need to agree on what we want to support specially about the 'nil' values.

**Comment by binhn** [comment]
@bcbrock That's not a use-case. I know and understand that generic key-value DBs support empty keys, but what use-case would a blockchain app use empty (or nil) keys? Fabric is an app of RocksDB, but that doesn't mean it has to support all the capabilities and features of RocksDB.

**Comment by bcbrock** [comment]
@binhn I gave a use case based on indirection of keys stored in a table:

_I've seen examples where keys are stored as fields of tables. If a field is "" it indicates a default value, which is represented by a key/value pair ""/<default>._

Is there some fundamental difficulty with allowing the empty string as a key?

**Comment by srderson** [comment]
Can we agree this PR is a step in the right direction?

@bcbrock I think @manish-sethi answered your question at https://github.com/hyperledger/fabric/pull/1471/files#r63202854. Empty keys are broken today, so we respond with an error. This does not prevent us from adding support in a future PR.

**Comment by binhn** [comment]
@bcbrock thanks for clarifying though I still see that as marginal use-case

@srderson I agreed to merge this pr and address empty key in a separate pr/issue

File: core/ledger/ledger_test.go [hunk 2]
```diff
@@ -902,3 +902,24 @@ func TestCopyState(t *testing.T) {
 	values, _ := l.GetStateMultipleKeys("chaincodeID2", []string{"key1", "key2", "key3"}, true)
 	testutil.AssertEquals(t, values, [][]byte{[]byte("value1"), []byte("value2"), []byte("value3")})
 }
+
+func TestLedgerEmptyArrayValue(t *testing.T) {
```
Comment (srderson) [inline core/ledger/ledger_test.go:906]: Could you add tests to confirm the error is returned on a empty key or a nil value?

File: core/ledger/ledger.go [hunk 3]
```diff
@@ -207,6 +207,9 @@ func (ledger *Ledger) GetStateRangeScanIterator(chaincodeID string, startKey str
...
+		return fmt.Errorf("A nil key or a nil value is not supported. key=%s, value=%#v", key, value)
```
Comment (srderson) [inline core/ledger/ledger.go]: This error should specify an empty key instead of nil.

File: core/ledger/ledger_test.go (hunk 2 above)
Comment (manish-sethi) [inline core/ledger/ledger_test.go:906]: Added a test for wrong input (last test in ledger_test.go) in commit aa93cf4

File: core/ledger/ledger.go [hunk 4]
```diff
@@ -37,12 +36,43 @@ import (
 
 var ledgerLogger = logging.MustGetLogger("ledger")
 
+//ErrorType represents the type of a ledger error
+type ErrorType string
+
+const (
+	//ErrorTypeWrongInput used to indicate the wrong input to ledger method
+	ErrorTypeWrongInput = ErrorType("WrongInput")
```
Comment (srderson) [inline core/ledger/ledger.go]: Minor issue, but can we use `InvalidArgument` instead of `WrongInput`?

Comment (manish-sethi) [inline core/ledger/ledger.go]: fixed in commit 2d272c8

**Comment by srderson** [comment]
LGTM. We will potentially address other issues (support for empty key and nil value) in a separate PRs.

File: integration/nwo/network.go [hunk 5]
```diff
@@ -413,6 +413,26 @@ func (n *Network) userCryptoDir(org *Organization, nodeOrganizationType, user, c
...
+func (n *Network) PeerOrgCADir(o *Organization) string {
+	return filepath.Join(
+		n.CryptoPath(),
+		"peerOrganizations",
+		o.Domain,
+		"ca",
+	)
+}
+
+func (n *Network) OrdererOrgCADir() string {
```
Comment (sykesm) [inline integration/nwo/network.go]: Nit: godoc

File: integration/nwo/network.go [hunk 6]
```diff
@@ -413,6 +413,26 @@ func (n *Network) userCryptoDir(org *Organization, nodeOrganizationType, user, c
 	)
 }
 
+// PeerOrgCADir returns the path to the folder containing the CA certificate(s) and/or
+// keys for the specified organization
```
Comment (sykesm) [inline integration/nwo/network.go]: nit: complete sentences in godoc

File: integration/pkcs11/p11key.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	"crypto/ecdsa"
+	"encoding/asn1"
+	"fmt"
+	"io"
+	"math/big"
+
+	"github.com/miekg/pkcs11"
+)
+
+// P11ECDSAKey test implementation of crypto.Signer
```
Comment (sykesm) [inline integration/pkcs11/p11key.go]: nit: throughout, any godoc should be complete sentences.

File: integration/pkcs11/p11key.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	privateKeyHandle pkcs11.ObjectHandle
+}
+
+// Public returns the corresponding public key for the
+// private key
+func (k *P11ECDSAKey) Public() crypto.PublicKey {
+	return k.publicKey
+}
+
+// Sign returns a crypto.Signer
```
Comment (sykesm) [inline integration/pkcs11/p11key.go]: This implements crypto.Signer; it doesn't return one.

File: integration/pkcs11/p11key.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+func (k *P11ECDSAKey) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	hash := opts.HashFunc()
+	hashLen := hash.Size()
+	if len(digest) != hashLen {
```
Comment (sykesm) [inline integration/pkcs11/p11key.go]: Why the temporary vars that are not referenced beyond this block instead of?
```go
if len(digest) != opts.HashFunc().Size() {
...
}
```

File: integration/pkcs11/pkcs11_suite_test.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+	. "github.com/onsi/gomega"
+)
+
+func TestPKCS11(t *testing.T) {
+	RegisterFailHandler(Fail)
+	lib, pin, label := bpkcs11.FindPKCS11Lib()
+	if lib != "" && pin != "" && label != "" {
+		RunSpecs(t, "PKCS11 Suite")
+	}
+}
```
Comment (sykesm) [inline integration/pkcs11/pkcs11_suite_test.go]: So, I think it's preferred to explicitly state that the suite is being skipped instead of just omitting the call to `RunSpecs`. Since this gets the `*testing.T` we can simply call `Skip(...)` on that.

So flip the test and call skip when any of the args is nil?

File: integration/pkcs11/pkcs11_test.go [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+				Library:  lib,
+			},
+		}
+
+		By("update bccsp peer config")
+		peerConfig := network.ReadPeerConfig(peer)
+		peerConfig.Peer.BCCSP = myBCCSP
+		network.WritePeerConfig(peer, peerConfig)
+
+		By("Update bccsp orderer config")
```
Comment (sykesm) [inline integration/pkcs11/pkcs11_test.go]: subject-verb agreement

File: integration/pkcs11/pkcs11_test.go [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+		ordererConfig := network.ReadOrdererConfig(orderer)
+		ordererConfig.General.BCCSP = myBCCSP
+		network.WriteOrdererConfig(orderer, ordererConfig)
+
+		network.Bootstrap()
+
+		ctx, sess := setupPKCS11Ctx(lib, label, pin)
+		defer ctx.CloseSession(sess)
+		Expect(err).NotTo(HaveOccurred())
```
Comment (sykesm) [inline integration/pkcs11/pkcs11_test.go]: The order of these are switched.

File: integration/pkcs11/pkcs11_test.go [hunk 13]
```diff
@@ -0,0 +1,411 @@
...
+		Eventually(process.Wait(), network.EventuallyTimeout).Should(Receive())
+
+		if network != nil {
+			network.Cleanup()
+		}
+		os.RemoveAll(tempDir)
+	})
+
+	It("Creates and join channel successfully", func() {
+
```
Comment (sykesm) [inline integration/pkcs11/pkcs11_test.go]: nit: blank lines

File: integration/pkcs11/pkcs11_test.go [hunk 14]
```diff
@@ -0,0 +1,411 @@
...
+		orderer := network.Orderer("orderer")
+		network.CreateAndJoinChannels(orderer)
+
+	})
+
```
Comment (sykesm) [inline integration/pkcs11/pkcs11_test.go]: nit: blank line

File: integration/pkcs11/pkcs11_test.go [hunk 15]
```diff
@@ -0,0 +1,411 @@
...
+})
+
+func configurePeerPKCS11(ctx *pkcs11.Ctx, sess pkcs11.SessionHandle, network *nwo.Network) {
+
+	peer := network.Peers[0]
```
Comment (sykesm) [inline integration/pkcs11/pkcs11_test.go]: Why just one peer?

File: integration/pkcs11/pkcs11_test.go [hunk 16]
```diff
@@ -0,0 +1,411 @@
...
+
+		orderer := network.Orderer("orderer")
+		network.CreateAndJoinChannels(orderer)
+
+	})
+
+})
+
+func configurePeerPKCS11(ctx *pkcs11.Ctx, sess pkcs11.SessionHandle, network *nwo.Network) {
+
```
Comment (sykesm) [inline integration/pkcs11/pkcs11_test.go:116]: nit: blank line

File: integration/pkcs11/pkcs11_test.go [hunk 17]
```diff
@@ -0,0 +1,411 @@
...
+	peer := network.Peers[0]
+	orgName := peer.Organization
+
+	peerPubKey, peerCSR, peerSerial := createCSR(ctx, sess, orgName, "peer")
+	adminPubKey, adminCSR, adminSerial := createCSR(ctx, sess, orgName, "admin")
+
+	// Retrieves org CA cert
+	orgCAPath := network.PeerOrgCADir(network.Organization(orgName))
+	caBytes, err := ioutil.ReadFile(filepath.Join(orgCAPath, "ca.org1.example.com-cert.pem"))
```
Comment (sykesm) [inline integration/pkcs11/pkcs11_test.go]: this is assuming that org1 and peer0 are together, isn't it?
Also, the organization domain should be resolved by calling `network.Organization(orgName)` and getting the `Domain` field.

File: integration/pkcs11/pkcs11_test.go [hunk 18]
```diff
@@ -0,0 +1,411 @@
...
+	Expect(err).NotTo(HaveOccurred())
+
+	By("Updating the peer signcerts")
+
```
Comment (sykesm) [inline integration/pkcs11/pkcs11_test.go:126]: nit: more blank lines.

(I'm cool with vertical whitespace between functional blocks of code but when there's a lot of it, it actually makes things harder to read.)

File: integration/pkcs11/pkcs11_test.go [hunk 19]
```diff
@@ -0,0 +1,411 @@
...
+	orgAdminMSPPath := network.PeerUserMSPDir(peer, "Admin")
+	updateMSPFolder(orgAdminMSPPath, "Admin@org1.example.com-cert.pem", newAdminPemCert)
+
+}
+
+func configureOrdererPKCS11(ctx *pkcs11.Ctx, sess pkcs11.SessionHandle, network *nwo.Network) {
+
+	orderer := network.Orderer("orderer")
+	orgName := "orderer"
```
Comment (sykesm) [inline integration/pkcs11/pkcs11_test.go]: the orderer org should be obtained from the org.
